xrmocap/model/mvp/projattn.py

```diff
@@ -0,0 +1,255 @@
+                input_spatial_shapes,
+                input_level_start_index,
+                input_padding_mask=None):
+        """Forward function for Projective Attention.
+
+        Args:
+            query :
+                (n_views, Length_{query}, C)
+            reference_points :
+                (n_views, Length_{query}, n_levels, 2),
+                range in [0, 1], top-left (0,0), bottom-right (1, 1),
+                including padding area or (n_views, Length_{query}, n_levels,
+                4), add additional (w, h) to form reference boxes
+            src_views :
+                list of (n_views, C, H, W), size n_levels,
+                [(n_views, C, H_0, W_0), (n_views, C, H_0, W_0), ...,
+                (n_views, C, H_{L-1}, W_{L-1})]
+            camera_ray_embeds :
+                Embedded camera rayts.
+            input_spatial_shapes :
+                (n_levels, 2), [(H_0, W_0), (H_1, W_1), ...,
+                (H_{L-1}, W_{L-1})]
+            input_level_start_index :
+                (n_levels, ), [0, H_0*W_0, H_0*W_0+H_1*W_1,
+                H_0*W_0+H_1*W_1+H_2*W_2, ...,
+                H_0*W_0+H_1*W_1+...+H_{L-1}*W_{L-1}]
+            input_padding_mask :
+                True for padding elements, False for non-padding elements.
+                Defaults to None.
+
+
+        Returns:
+            output : (n_views, Length_{query}, C)
+        """
+
+        n_views, Len_q, c = query.shape
+        feat_lvls = len(src_views)
+        sample_grid = torch.clamp(reference_points * 2.0 - 1.0, -1.1, 1.1)
+        ref_point_feat_views_alllvs = []
+
+        if 'use_rayconv' in self.projattn_pos_embed_mode or \
+                'use_2d_coordconv' in self.projattn_pos_embed_mode:
+            for lvl in range(feat_lvls):
+                ref_point_feat_views_alllvs.append(
+                    F.grid_sample(src_views[lvl],
+                                  sample_grid[:, :, lvl:lvl +
+                                              1, :]).squeeze(-1).permute(
+                                                  0, 2, 1))
+            input_flatten = torch.cat([
+                torch.cat([src.flatten(2)
+                           for src in src_views], dim=-1).permute(0, 2, 1),
+                torch.cat([cam.flatten(1, 2) for cam in camera_ray_embeds],
+                          dim=1)
+            ],
+                                      dim=-1)
+
+        elif self.projattn_pos_embed_mode == 'ablation_not_use_rayconv':
+            for lvl in range(feat_lvls):
+                ref_point_feat_views_alllvs.append(
+                    F.grid_sample(src_views[lvl],
+                                  sample_grid[:, :, lvl:lvl +
+                                              1, :]).squeeze(-1).permute(
+                                                  0, 2, 1))
+            input_flatten = torch.cat([src.flatten(2) for src in src_views],
+                                      dim=-1).permute(0, 2, 1)
+
```

Comment:
Shall we get an error when `self.projattn_pos_embed_mode` is another type?

Reply:
Added in [ad43f3b](https://github.com/openxrlab/xrmocap/pull/21/commits/ad43f3b61eb55fb3c7998d07ae3e7b87c871347f), logger will log the error if unrecognized mode is given.